### PR TITLE
[MDB-29707] Metatdata cleaner with system drop replica

### PR DESCRIPTION
**Pull request by MikhailBurdukov**
A little performance test:
```
500 replicated tables + 0 ReplicatedDB
    old 5 runs: 
        median:  0m44.959s
        list : [2m56.825s, 2m56.635s, 0m44.959s, 0m41.463s, 0m44.481s]
    new 5 runs:  
        median:  0m9.965s
        list: [0m6.865s, 0m9.965s, 0m10.014s, 0m6.701s, 0m10.085s]


500 replicated tables + 200 Replicated databases
    old 5 runs: 
        median:  3m39.932s
        list : [0m50.593s, 0m52.412s, 3m39.697s, 3m40.830s, 3m39.932s]
    new 5 runs:  
        median:  0m8.550s
        list: [0m13.685s, 0m13.335s, 0m8.550s, 0m8.320s, 0m8.627s]
```

gen script:
```
REPLICA="non_exists"
DB='tests_db'

clickhouse client --query "ATTACH DATABASE $DB"
clickhouse client --query "DROP DATABASE $DB"

clickhouse client --query "CREATE DATABASE $DB"


for i in {1..500}
do
   QUERY="CREATE TABLE $DB.t$i (a int) ENGINE=ReplicatedMergeTree('/clickhouse/tables/$DB/t$i','$REPLICA') ORDER BY a"

   echo "$QUERY"
   clickhouse client --query "$QUERY"
done

clickhouse client --query "DETACH DATABASE $DB"


for i in {1..200}
do
        DB_REPL="repl$i"
        clickhouse client --query "ATTACH DATABASE $DB_REPL"
        clickhouse client --query "DROP DATABASE IF EXISTS $DB_REPL"
        QUERY="CREATE DATABASE $DB_REPL ENGINE = Replicated('/clickhouse/databases/$DB_REPL', 'shard1', '$REPLICA')"
        echo $QUERY
        clickhouse client --query "$QUERY"
        clickhouse client --query "DETACH DATABASE $DB_REPL"
done
```

### CODE REVIEW

File: ch_tools/chadmin/internal/zookeeper.py [hunk 1]
```diff
@@ -255,6 +263,129 @@ def wrapper(ctx, nodes, *args, **kwargs):
...
+            async_task = self._zk_client.get_children_async(path)
+            self._queue_active.append((path, async_task))
+
+    def _update_active_queue(self, block_until_finised_tasks: bool = False) -> None:
+        """
+        Moves executed tasks from active to finished queues.
+        """
+
+        if block_until_finised_tasks:
+            self._queue_active[0][1].wait()
```
Comment (aalexfvk) [inline ch_tools/chadmin/internal/zookeeper.py:314]: `self._queue_active` can be empty

File: ch_tools/chadmin/internal/zookeeper.py [hunk 2]
```diff
@@ -255,6 +263,129 @@ def wrapper(ctx, nodes, *args, **kwargs):
...
+        add() -> add list nodes to get children.
+        get() -> get the one node with list of children.
+    """
+
+    def __init__(self, zk: KazooClient, max_parrallel_tasks: int = 5):
+        self._zk_client = zk
+        self._max_active_tasks = max_parrallel_tasks
+
+        self._queue_pending: Deque[str] = deque()
+        self._queue_active: Deque[Tuple[str, IAsyncResult]] = deque(
```
Comment (aalexfvk) [inline ch_tools/chadmin/internal/zookeeper.py:289]: Let's make dataclass for the tuple to get rid of naked indices like `[0][1]`

File: ch_tools/chadmin/internal/zookeeper.py [hunk 3]
```diff
@@ -255,6 +263,129 @@ def wrapper(ctx, nodes, *args, **kwargs):
...
+        without_block - get one finished task if there are no finished, calls itself with block.
+        """
+        self._update_queues(with_block)
+
+        if len(self._queue_finished):
+            return self._queue_finished.popleft()
+
+        if not self._exists_tasks_to_do():
+            return None
+
```
Comment (aalexfvk) [inline ch_tools/chadmin/internal/zookeeper.py:362]: It looks like we can add waiting for the first ready result here and remove `with_block` and `block_until_finished_tasks` parameters from everywhere

File: ch_tools/chadmin/internal/zookeeper.py [hunk 4]
```diff
@@ -271,119 +402,178 @@ def clean_zk_metadata_for_hosts(
...
 
-    def _set_replicas_is_lost(zk, table_paths, nodes):
-        """
-        Set flag <path>/replicas/<replica_name>/is_lost to 1
-        """
-        for path in table_paths:
-            replica_path = os.path.join(path, "replicas")
-            if not zk.exists(replica_path):
+        for replicated_object in replicated_objects:
+            # Actually rn in the ch (10.24) there are no secure way to determine that node is the root of replicated table.
```
Comment (aalexfvk) [inline ch_tools/chadmin/internal/zookeeper.py:468]:
```suggestion
            # Actually there is no a reliable way to determine that node is the root of replicated table in the CH (24.10)
```

File: ch_tools/chadmin/internal/zookeeper.py [hunk 5]
```diff
@@ -271,119 +402,178 @@ def clean_zk_metadata_for_hosts(
...
+            # So we are accuming that if object not database then it is table.
```
Comment (aalexfvk) [inline ch_tools/chadmin/internal/zookeeper.py:469]:
```suggestion
            # So we are assuming that if object is not a database then it is a table.
```

File: ch_tools/chadmin/internal/zookeeper.py [hunk 6]
```diff
@@ -271,119 +402,178 @@ def clean_zk_metadata_for_hosts(
...
+            #
+            # Replicated table
+            # https://github.com/ClickHouse/ClickHouse/blob/eb984db51ea0309dd607eaf98280315b42347f65/src/Interpreters/InterpreterSystemQuery.cpp#L1029
+            #
+            # Replicated Database
+            # https://github.com/ClickHouse/ClickHouse/blob/eb984db51ea0309dd607eaf98280315b42347f65/src/Databases/DatabaseReplicated.cpp#L1529
+            try:
+                is_database = bool(
+                    zk.get(replicated_object.path)[0] == REPLICATED_DATABASE_MARKER
+                )
```
Comment (aalexfvk) [inline ch_tools/chadmin/internal/zookeeper.py:479]: Why is casting to `bool` needed here ?

File: ch_tools/chadmin/internal/zookeeper.py [hunk 7]
```diff
@@ -271,119 +402,178 @@ def clean_zk_metadata_for_hosts(
...
+            replicas_list = _get_children(
+                zk, os.path.join(replicated_object.path, "replicas")
+            )
+
+            for replica in replicas_list:
+                if is_database:
+                    shard, replica_parsed = replica.split("|")
+                    if replica_parsed in nodes_set:
+                        if replicated_object.path not in databases_to_cleanup:
+                            databases_to_cleanup[replicated_object.path] = []
```
Comment (aalexfvk) [inline ch_tools/chadmin/internal/zookeeper.py:501]: How about using `defaultdict` for the `databases_to_cleanup` ?

File: ch_tools/common/config.py [hunk 8]
```diff
@@ -37,6 +37,7 @@
         "unfreeze_timeout": 10 * 60,
         "restart_replica_timeout": 10 * 60,
         "restore_replica_timeout": 10 * 60,
+        "drop_replica_timeout": 10 * 60,
```
[the remark this answers is not in the document]
Comment (aalexfvk) [reply ch_tools/common/config.py:40]: I think its ok. We will adjust it if necessary

File: ch_tools/chadmin/internal/zookeeper.py (hunk 3 above)
Comment (MikhailBurdukov) [reply ch_tools/chadmin/internal/zookeeper.py:373]: Yes, we can. But this will lead to unnecessary waiting every time we call the queue update.

Comment (aalexfvk) [reply ch_tools/chadmin/internal/zookeeper.py:373]: What if we remove waiting from `_update_active_queue`  at all  and insert it here ?

Comment (MikhailBurdukov) [reply ch_tools/chadmin/internal/zookeeper.py]: moved implementation with wait to the separate function

**Review by aalexfvk — APPROVED** [review]

**Pull request merged**